redis/db/redisbp/prometheus.go

```diff
@@ -0,0 +1,133 @@
+package redisbp
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	promNamespace = "redisbp"
+	subsystemPool = "pool"
+
+	nameLabel = "pool"
+)
+
+// exporter provides an interface for Prometheus metrics.
+type exporter struct {
+	client PoolStatser
+	name   string
+
+	poolHitsCounterDesc     *prometheus.Desc
+	poolMissesCounterDesc   *prometheus.Desc
+	poolTimeoutsCounterDesc *prometheus.Desc
+	totalConnectionsDesc    *prometheus.Desc
+	idleConnectionsDesc     *prometheus.Desc
+	staleConnectionsDesc    *prometheus.Desc
+}
+
+func newExporter(client PoolStatser, name string) *exporter {
+	labels := []string{
+		nameLabel,
+	}
+
+	return &exporter{
+		client: client,
+		name:   name,
+
+		// Upstream docs: https://pkg.go.dev/github.com/go-redis/redis/v8/internal/pool#Stats
+
+		// Counters.
+		poolHitsCounterDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(promNamespace, subsystemPool, "hits_total"),
+			"Number of times free connection was found in the pool",
+			labels,
+			nil,
+		),
+		poolMissesCounterDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(promNamespace, subsystemPool, "misses_total"),
+			"Number of times free connection was NOT found in the pool",
+			labels,
+			nil,
+		),
+		poolTimeoutsCounterDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(promNamespace, subsystemPool, "timeouts_total"),
+			"Number of times a wait timeout occurred",
+			labels,
+			nil,
+		),
+
+		// Gauges.
+		totalConnectionsDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(promNamespace, subsystemPool, "connections"),
+			"Number of connections in this redisbp pool",
+			labels,
+			nil,
+		),
+		idleConnectionsDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(promNamespace, subsystemPool, "idle_connections"),
+			"Number of idle connections in this redisbp pool",
+			labels,
+			nil,
+		),
+		staleConnectionsDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(promNamespace, subsystemPool, "stale_connections"),
+			"Number of stale connections in this redisbp pool",
+			labels,
+			nil,
+		),
+	}
+}
+
+// Describe implements the prometheus.Collector interface.
+func (e *exporter) Describe(ch chan<- *prometheus.Desc) {
```

Comment:
🔕 In this approach, will it cause problems if someone call `NewMonitoredClient` twice (and thus create two `exporter`s to be registered)? maybe we should add a test to make sure that is fine?

Reply:
actually it will cause the second `Register` call to return an error, as prometheus consider them duplicated: https://github.com/prometheus/client_golang/blob/v1.12.1/prometheus/registry.go#L295-L299

so we need to go back to the no-op `Describe` and dynamically register it way.